### ISSUE TITLE
Feature request: add extfs quota support for overlay(2)

### ISSUE THREAD

**Issue by tswift242** [docker/docker#29364]
**Description**

This is a request to add quota support for ext4 for the overlay and overlay2 storage drivers, similar to what has been done in https://github.com/docker/docker/pull/24771 and https://github.com/docker/docker/pull/24807.

https://github.com/docker/docker/pull/24771#issuecomment-250274944 mentions that ext4 quota support was added in kernel 4.5. The next LTS Linux kernel (4.9) was just released this past Sunday, which will have this support, so maybe it's worth looking into this again.

**Comment by justincormack** [comment]
Yes, all the next lot of distros will be 4.9 based, so this seems timely.

**Comment by xianwei2016** [comment]
We are happy to see this feature can be supported soon.

**Comment by xianwei2016** [comment]
The kernel commit for ext4fs project quota support is in: https://kernelnewbies.org/Linux_4.5

One key question on this feature: does this support ext4fs directory level quota on per container? not quota based on user/group or block level.

What we really want is like below:
docker run -l DiskQuota=60g ...
docker run -l DiskQuota=/=60g;/data1=50g ...

**Comment by xianwei2016** [comment]
@justincormack , @cpuguy83

**Comment by xianwei2016** [comment]
Hi @tswift242, is this means we can set extfs disk quota based on directory level just as the example I posted? Thanks very much.

**Comment by xiaods** [comment]
any summary update?

**Comment by thaJeztah** [comment]
There's an open PR to implement this; https://github.com/moby/moby/pull/34364, but it looks like it has stalled; let me see if we can have a look at that one again

**Comment by katsar0v** [comment]
Is there any update on this?

**Comment by ryanwohara** [comment]
It would be amazing to see progress on this in 2019 🙏

**Comment by katsar0v** [comment]
Yep, +1 on this. This feature is requested for long time.

**Comment by xiaods** [comment]
It would be amazing to see progress on this in 2019 🙏

**Comment by mhemrg** [comment]
+1

**Comment by supernovae** [comment]
+1

**Comment by cchen-flexport** [comment]
+1

**Comment by naioja** [comment]
+1

**Comment by kimtiago** [comment]
+1

**Comment by MohamedKarrab** [comment]
come on guys we need this on extfs

**Comment by Vann-Dev** [comment]
Up 🆙

**Comment by rajat709** [comment]
+1

**Comment by ROMYIM** [comment]
+1

**Comment by acheroncrypto** [comment]
https://github.com/moby/moby/blob/72752cf5108842bef17616eef4b386cc4a404be7/daemon/graphdriver/overlay2/overlay.go#L329

Why is this still not supported 10 years later? Setting storage limits is an essential part of any sort of sandboxing. `extfs` is also much more common than `xfs`.

**Comment by ell1e** [comment]
I wonder, is btrfs supported for this? But yea, ext4 support would be pretty useful too.
